devito/passes/iet/mpi.py

```diff
@@ -59,120 +59,116 @@ def _hoist_halospots(iet):
-                            break
-                    else:
-                        hsmapper[hs] = hsmapper[hs].drop(f)
-                        imapper[i].append(hs.halo_scheme.project(f))
-                        break
+                # Compare hs0 to subsequent halo_spots, looking for optimization
+                # possibilities
+                _process_halo_to_halo(hs0, hs1, iters, scopes, hsmapper, imapper)
 
-    # Post-process analysis
     mapper = {i: HaloSpot(i._rebuild(), HaloScheme.union(hss))
               for i, hss in imapper.items()}
     mapper.update({i: i.body if hs.is_void else i._rebuild(halo_scheme=hs)
                    for i, hs in hsmapper.items()})
 
-    # Transform the IET hoisting/dropping HaloSpots as according to the analysis
-    iet = Transformer(mapper, nested=True).visit(iet)
-
-    # Clean up: de-nest HaloSpots if necessary
-    mapper = {}
-    for hs in FindNodes(HaloSpot).visit(iet):
-        if hs.body.is_HaloSpot:
-            halo_scheme = HaloScheme.union([hs.halo_scheme, hs.body.halo_scheme])
-            mapper[hs] = hs._rebuild(halo_scheme=halo_scheme, body=hs.body.body)
     iet = Transformer(mapper, nested=True).visit(iet)
 
     return iet
 
 
-def _merge_halospots(iet):
-    """
-    Merge HaloSpots on the same Iteration tree level where all data dependencies
-    would be honored.
-    """
+def _process_halo_to_halo(hs0, hs1, iters, scopes, hsmapper, imapper):
 
-    # Merge rules -- if the retval is True, then it means the input `dep` is not
-    # a stopper to halo merging
+    # Loop over the functions in the HaloSpots
+    for f, v in hs1.fmapper.items():
+        # If no time accesses, skip
+        if not hs1.halo_scheme.fmapper[f].loc_indices:
+            continue
 
-    def rule0(dep, hs, loc_indices):
-        # E.g., `dep=W<f,[t1, x]> -> R<f,[t0, x-1]>` => True
-        return not any(d in hs.dimensions or dep.distance_mapper[d] is S.Infinity
-                       for d in dep.cause)
+        # If the function is not in both HaloSpots, skip
+        if (*hs0.functions, *hs1.functions).count(f) < 2:
+            continue
 
-    def rule1(dep, hs, loc_indices):
-        # TODO This is apparently never hit, but feeling uncomfortable to remove it
-        return (dep.is_regular and
-                dep.read is not None and
-                all(not any(dep.read.touched_halo(d.root)) for d in dep.cause))
+        for iter in iters:
+            # If also merge-able we can start hoisting the latter
+            for dep in scopes[iter].d_flow.project(f):
+                if not any(r(dep, hs1, v.loc_indices) for r in merge_rules()):
+                    break
+            else:
+                hse = hs1.halo_scheme.fmapper[f]
+                raw_loc_indices = {}
+                # Entering here means we can lift, and we need to update
+                # the loc_indices with known values
+                # TODO: Can I get this in a more elegant way?
```

Comment:
not only that, you should also check the iteration direction

Reply:
which means, btw, you have to add a pass for a gradient-like operator (as long as time flows backwards, we're good)

Reply:
not sure if it's useful or not, but there are various places in which we have to distinguish our actions depending on whether an iteration flows forward or backwards. For example:

https://github.com/devitocodes/devito/blob/master/devito/passes/clusters/asynchrony.py#L231
https://github.com/devitocodes/devito/blob/master/devito/mpi/halo_scheme.py#L533

and a few others (some of these are quite legacy unfortunately)

I wonder whether there's an utility function that we can write and use in more than one place, but I'm far from sure, as each of these *seem* very special cases... but it's worth looking into it

Reply:
I though the rules there would ensure that we do not hoist some wrong t-slice.
Shouldn;t we just assert:
assert hs0.loc_dirs == hs1.loc_dirs
in order to do any of those transforms?

Reply:
done